From 95ae070e409f5acaca0eaf90a8b8e69631d68d69 Mon Sep 17 00:00:00 2001
From: Colin Patrick Mccabe <cmccabe@cloudera.com>
Date: Mon, 20 Oct 2014 18:24:53 -0700
Subject: [PATCH 208/596] HDFS-7266. HDFS Peercache enabled check should not
 lock on object (awang via cmccabe)

(cherry picked from commit 4799570dfdb7987c2ac39716143341e9a3d9b7d2)
(cherry picked from commit fdd7c7145e739767ab4f337e16f3ef05d09e4ff1)
---
 .../java/org/apache/hadoop/hdfs/PeerCache.java     |   12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/PeerCache.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/PeerCache.java
index fbe4e15..07c562e 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/PeerCache.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/PeerCache.java
@@ -140,12 +140,15 @@ public String toString() {
    * @return             An open Peer connected to the DN, or null if none
    *                     was found. 
    */
-  public synchronized Peer get(DatanodeID dnId, boolean isDomain) {
+  public Peer get(DatanodeID dnId, boolean isDomain) {
 
     if (capacity <= 0) { // disabled
       return null;
     }
+    return getInternal(dnId, isDomain);
+  }
 
+  private synchronized Peer getInternal(DatanodeID dnId, boolean isDomain) {
     List<Value> sockStreamList = multimap.get(new Key(dnId, isDomain));
     if (sockStreamList == null) {
       return null;
@@ -174,7 +177,7 @@ public synchronized Peer get(DatanodeID dnId, boolean isDomain) {
   /**
    * Give an unused socket to the cache.
    */
-  public synchronized void put(DatanodeID dnId, Peer peer) {
+  public void put(DatanodeID dnId, Peer peer) {
     Preconditions.checkNotNull(dnId);
     Preconditions.checkNotNull(peer);
     if (peer.isClosed()) return;
@@ -183,7 +186,10 @@ public synchronized void put(DatanodeID dnId, Peer peer) {
       IOUtils.cleanup(LOG, peer);
       return;
     }
- 
+    putInternal(dnId, peer);
+  }
+
+  private synchronized void putInternal(DatanodeID dnId, Peer peer) {
     startExpiryDaemon();
 
     if (capacity == multimap.size()) {
-- 
1.7.9.5

